crates/lib3h/src/rrdht_util/r_value.rs

```diff
@@ -0,0 +1,335 @@
+/// Also, this is a bit of a naive first implementation, expect the specific
+/// algorithm / heuristics to be updated on an on-going basis.
+pub fn get_recommended_storage_arc_radius(
+    peer_record_set: &RValuePeerRecordSet,
+    target_minimum_r_value: f64,
+    target_maximum_r_value: f64,
+    current_arc_radius: Option<u32>,
+) -> u32 {
+    let pct_of_space_covered: f64 =
+        peer_record_set.arc_of_included_peer_records.length() as f64 / ARC_LENGTH_MAX as f64;
+    let interp_total_node_count =
+        peer_record_set.peer_records.len() as f64 * (1.0 / pct_of_space_covered);
+    if interp_total_node_count < target_minimum_r_value * 2.0 {
+        // consider this network immature, recommend full coverage
+        return ARC_RADIUS_MAX;
+    }
+
+    let cur_r_value = interpolate_r_value_for_given_arc(peer_record_set);
+
+    let mut count: u64 = 0;
+    let mut rad_avg: f64 = 0.0;
+
+    #[allow(clippy::explicit_counter_loop)]
+    for record in peer_record_set.peer_records.iter() {
+        count += 1;
+
+        let rad = record.storage_arc.radius();
+
+        rad_avg = (rad_avg * (count as f64 - 1.0) + f64::from(rad)) / count as f64;
+    }
+
+    let out_radius: f64 = if cur_r_value < target_minimum_r_value {
+        // our network is unhealthy! let's try to capture more!
+
+        // let's double the current average for our target mid-point
+        let new_mid_rad = rad_avg * 2.0;
+
+        match current_arc_radius {
+            None => new_mid_rad,
+            Some(cur) => {
+                let cur = f64::from(cur);
+
+                // try to maintain some stability
+                // we'll accept any current from 1.5 to 3 times the average
+                let new_min_rad = rad_avg * 1.5;
+                let new_max_rad = rad_avg * 3.0;
+
+                if cur > new_min_rad && cur < new_max_rad {
+                    cur
+                } else {
+                    new_mid_rad
+                }
+            }
+        }
+    } else if cur_r_value > target_maximum_r_value {
+        // our network is heavy! let's pull back
+
+        // let's shoot for somewhere between half and 3/4 the current average
+        let new_mid_rad = rad_avg * 0.625;
+        match current_arc_radius {
+            None => new_mid_rad,
+            Some(cur) => {
+                let cur = f64::from(cur);
+
+                // try to maintain some stability
+                let new_min_rad = rad_avg * 0.5;
+                let new_max_rad = rad_avg * 0.75;
+
+                if cur > new_min_rad && cur < new_max_rad {
+                    cur
+                } else {
+                    new_mid_rad
+                }
+            }
+        }
+    } else {
+        // our network is perfect, let's try to keep it this way!
+        match current_arc_radius {
+            None => rad_avg,
+            Some(cur) => f64::from(cur),
+        }
+    };
+
+    // we might be tempted to correct for any shortage in our own uptime here
```

Comment:
I sense a correction factor that incorporates uptime could be useful as a direct correlation (rather than inverse). Nodes that are up within the DHT for longer should be storing more... they have more time to accumulate data, they're more invested in the DHT, and if they're returning, they probably already have more DHT data so gossip to update them to current will require less bandwidth. This could be really useful for ensuring new peer entrants don't inundate the DHT with gossip update requests, too.

Reply:
> Nodes that are up within the DHT for longer should be storing more

Only on the time-scale of minutes... hApps that require hours of downloads should hopefully be outliers... I think for the most part, you can assume everyone on the network holds as much data as they plan to.

.. I'm not sure I understand the other bits you mention here... All I was saying is that the "R" value target we calculated already adjusts for uptime - so if we increased the recommendation to account for our own uptime, everybody would slowly be increasing their arcs for no reason.

Reply:
Let's come back to discussing this if we find DHT entry gossip gets to be too intensive. I sense it would be a good way to gracefully serve a large number of new entrants or large number of very infrequent users.